### PR TITLE
Add common object package

### PR DESCRIPTION
**Pull request by xyhuang**
This adds a common package with definitions of config objects and utility functions for internal use. This makes it easy to edit and use schema-bound objects and generate schema from code in the future.

part of #66

### CODE REVIEW

**Comment by sergey-serebryakov** [comment]
Hi @xyhuang, thanks!
It convenient to semi-automatically instantiate configuration objects based upon ML schema files. Couple things I found looking at the source files:

- Should it be `return NotImplemented` or `raise NotImplementedError` in [BaseField](https://github.com/xyhuang/mlbox/blob/runner-commons/mlcommons_box/mlcommons_box/common/objects/base.py) class.
- There's a typo in [base.py:97](https://github.com/xyhuang/mlbox/blob/runner-commons/mlcommons_box/mlcommons_box/common/objects/base.py#L97) - should be 'StandardObject'.
- [DictOfObject::from_primitive](https://github.com/xyhuang/mlbox/blob/runner-commons/mlcommons_box/mlcommons_box/common/objects/base.py#L190) does not return anything (should it be `return self`?).
- In [ListOfObjects::validate](https://github.com/xyhuang/mlbox/blob/runner-commons/mlcommons_box/mlcommons_box/common/objects/base.py#L135): `item_class` -> `ListOfObject.item_class`. Same for   
 [DictOfObject::validate](https://github.com/xyhuang/mlbox/blob/runner-commons/mlcommons_box/mlcommons_box/common/objects/base.py#L135): `value_class` -> `DictOfObject.value_class`.

This is kind of a discussion topic. Being able to specify configuration parameters via `fields` variable is convenient, but it prevents some of the IDEs to show available object fields when a developer types object_instance_variable and then `.`:
```
cfg = PlatformConfig()
cfg.   # IDE does not show platform and container
```

It can be done in latest Python versions by providing type annotations at class level:
```
class PlatformConfig(base.StandardObject):
    platform: base.ObjectField
    container: base.ObjectField
```
Not sure if it's the right way though.

**Comment by sub-mod** [comment]
@xyhuang thank, can you also rebase the commits to a single commit.

File: mlcommons_box/mlcommons_box/common/objects/base.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+import logging
+import os
+import typing
+
+import mlspeclib
+
+logger = logging.getLogger(__name__)
+
+
+class BaseField(abc.ABC):
```
Comment (sub-mod) [inline mlcommons_box/mlcommons_box/common/objects/base.py:12]: what is abc.ABC ?

**Comment by xyhuang** [comment]
@sergey-serebryakov 

> it prevents some of the IDEs to show available object fields

this is a good question. If we do the annotations, do we need to statically initialize them in __init__? In that case maybe we can make the improvements in a follow up patch as it seems to need some more changes than just the annotations.

File: mlcommons_box/mlcommons_box/common/objects/base.py (hunk 1 above)
Comment (xyhuang) [reply mlcommons_box/mlcommons_box/common/objects/base.py:12]: this is the "abstract base class" that allows us to define abstract classes/methods

**Comment by sergey-serebryakov** [comment]
@xyhuang I need to read more about that, but my impression is that annotations at class level can be used for dynamic attributes. Those, declared in `__init__` are not an issue.
Yeah, we can improve this subsequent PRs.

**Pull request merged**